devito/types/separable.py

```diff
@@ -0,0 +1,195 @@
+        Type of separation as a product or sum of Function, 'sum' or 'prod'.
+    separated : Dimension or tuple of Dimensions
+        Dimension in which the Function is separable.
+
+    Examples
+    --------
+    Creation
+
+    >>> from devito import Grid, TimeFunction
+    >>> grid = Grid(shape=(4, 4, 5))
+    >>> x, y, z = grid.dimensions
+    >>> SeparableFunction(name='a', grid=grid)
+    a_x(x)*a_y(y)*a_z(z)
+    >>> SeparableFunction(name='b', grid=grid, op='sum')
+    b_x(x) + b_y(y) + b_z(z)
+    >>> SeparableFunction(name='c', grid=grid, op='sum', separated=x)
+    c_x(x) + c_yz(y, z)
+    """
+    _spearation_op = {'sum': Add, 'prod': Mul}
+
+    def __init__(self, *args, **kwargs):
+        if not self._cached():
+            # Is it sum or product separable
+            self._sep_op = self._spearation_op[kwargs.get('op', 'prod')]
+
+            # Conventional Function inputs
+            self._grid = kwargs.get('grid')
+            self._space_order = kwargs.get('space_order', 1)
+
+            # Initialize subfunctions
+            # Separable dimensions
+            sep_d, nonsep_d = self.__setup_dimensions__(**kwargs)
+            self._dims = sep_d + (nonsep_d,)
+
+            func_list = self.__setup_functions__(sep_d, nonsep_d, **kwargs)
+
+            expr = self._sep_op(*func_list)
+
+            self._subfunctions = {d: f for d, f in zip(self.dims, func_list)}
+            self._args = (expr,)
+            self._fd = self.expr._fd
+
+    def __setup_functions__(self, sep_d, nonsep_d, **kwargs):
+
+        func_list = [self.function(d, **kwargs) for d in sep_d]
+        if len(nonsep_d) > 0:
+            func_list.append(self.function(nonsep_d, **kwargs))
+
+        return func_list
+
+    def function(self, dimensions, **kwargs):
+        dims = as_tuple(dimensions)
+        func_type = Function
+        names = "".join(d.name for d in dims)
+        shape = tuple(self.grid.dimension_map[d].loc for d in dims)
+        kwargs["name"] = '%s_%s' % (self.name, names)
+        func = func_type(dimensions=dims, shape=shape, **kwargs)
+        return func
+
+    @property
+    def sep_op(self):
+        return self._sep_op
+
+    @property
+    def grid(self):
+        return self._grid
+
+    @property
+    def space_order(self):
+        return self._space_order
+
+    def __setup_dimensions__(self, **kwargs):
+        sep = as_tuple(kwargs.get('separated', self.grid.dimensions))
+        nonsep = tuple(d for d in self.grid.dimensions if d not in sep)
+        return sep, nonsep
+
+    @classmethod
+    def __dtype_setup__(cls, **kwargs):
+        return kwargs.get('grid').dtype
+
+    def __str__(self):
+        return self.expr.__repr__()
+
+    __repr__ = __str__
+
+    @property
+    def _latex(self):
+        return self.expr._latex
+
+    @property
+    def evaluate(self):
+        return self.expr.evaluate
+
+    @property
+    def expr(self):
+        return self.args[0]
+
+    @property
+    def dims(self):
```

Comment:
this doesn't have any sense...users expect ``dimensions`` and ``indices``, not ``dims``. This is a result of not subclassing the right thing